runtime/runc_jailer.go

```diff
@@ -0,0 +1,370 @@
+				filename := filepath.Base(v.Path)
+				v.Path = filepath.Join("/", filename)
+				m.Cfg.VsockDevices[i] = v
+			}
+
+			if err := filepath.Walk(rootPath, func(name string, info os.FileInfo, err error) error {
+				if err != nil {
+					return err
+				}
+
+				j.logger.WithField("path", name).Debug("chowning")
+
+				return os.Chown(name, int(j.uid), int(j.gid))
+			}); err != nil {
+				return errors.Wrap(err, "failed to chown directories")
+			}
+
+			j.logger.Info("Successfully ran jailer handler")
+			return nil
+		},
+	}
+}
+
+// BuildLinkFifoHandler will return a new firecracker.Handler with the function
+// that will allow linking of the fifos making them visible to Firecracker.
+func (j runcJailer) BuildLinkFifoHandler() firecracker.Handler {
+	return firecracker.Handler{
+		Name: jailerFifoHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			contentsPath := j.RootPath()
+			fifoFileName := filepath.Base(m.Cfg.LogFifo)
+			newFifoPath := filepath.Join(contentsPath, fifoFileName)
+			if err := os.Link(m.Cfg.LogFifo, newFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.LogFifo = newFifoPath
+
+			metricFifoFileName := filepath.Base(m.Cfg.MetricsFifo)
+			newMetricFifoPath := filepath.Join(contentsPath, metricFifoFileName)
+			if err := os.Link(m.Cfg.MetricsFifo, newMetricFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.MetricsFifo = newMetricFifoPath
+
+			return nil
+		},
+	}
+}
+
+// ExposeDeviceToJail will inspect the given file, srcFilePath, and based on the
+// file type, proper handling will occur to ensure that the file is visible in
+// the jail.
+func (j runcJailer) ExposeDeviceToJail(srcMountPath string) error {
+	uid := j.uid
+	gid := j.gid
+
+	stat := syscall.Stat_t{}
+	if err := syscall.Stat(srcMountPath, &stat); err != nil {
+		return err
+	}
+
+	path := j.RootPath()
+	dst := ""
+	if (stat.Mode & syscall.S_IFMT) == syscall.S_IFBLK {
+		path = filepath.Join(path, "dev")
+		dst = filepath.Join(path, filepath.Base(srcMountPath))
+		if err := exposeBlockDeviceToJail(dst, int(stat.Rdev), int(uid), int(gid)); err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("unsupported mode: %v", stat.Mode)
+	}
+
+	return nil
+}
+
+// exposeBlockDeviceToJail will call mknod on the block device to ensure
+// visibility of the device
+func exposeBlockDeviceToJail(dst string, rdev, uid, gid int) error {
+	if err := syscall.Mknod(dst, syscall.S_IFBLK, rdev); err != nil {
+		return err
+	}
+
+	if err := os.Chmod(dst, 0600); err != nil {
+		return err
+	}
+
+	if err := os.Chown(dst, uid, gid); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func copyFile(src, dst string, mode os.FileMode) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", src)
+	}
+	defer srcFile.Close()
+
+	info, err := os.Stat(dst)
+	switch {
+	case err == nil:
+		// no error indicates the path already exists, we check to see if it is a
```

Comment:
Do we really want to have this logic here?  I think it would make sense to require a caller to pass the full destination instead of just a directory.